api/Updated-Metrics.md

```diff
@@ -0,0 +1,368 @@
+    },
+    {
+        'sample_weight': [1,2,1,1,3, ... ],
+        'pos_label' = 'Granted'
+    }
+]
+```
+If users had a lot of functions with a lot of custom arguments, this could get error-prone and difficult to debug.
+
+## User Conveniences
+
+We can consider allowing the metric function given to `group_summary()` to be represented by a string.
+We would provide a mapping of strings to suitable functions.
+This would make the following all equivalent:
+```python
+>>> r1 = group_summary(sklearn.accuracy_score, y_true, y_pred, sensitive_features=A_sex)
+>>> r2 = group_summary('accuracy_score', y_true, y_pred, sensitive_features=A_sex)
+>>> r3 = accuracy_score_group_summary( y_true, y_pred, sensitive_features=A_sex)
+```
+We would also allow mixtures of strings and functions in the multiple metric case.
+
+## Generality
+
+Throughout this document, we have been describing the case of classification metrics.
+However, we do not actually require this.
+It is the underlying metric function which gives meaning to the `y_true` and `y_pred` lists.
+So long as these are of equal length (and equal in length to the sensitive feature list - which _will_ be treated as a categorical), then `group_summary()` does not actually care about their datatypes.
+For example, each entry in `y_pred` could be a dictionary of predicted classes and accompanying probabilities.
+Or the user might be working on a regression problem, and both `y_true` and `y_pred` would be floating point numbers (or `y_pred` might even be a tuple of predicted value and error).
+So long as the underlying metric understands the datastructures, `group_summary()` will not care.
+
+There will be an effect on the `GroupedMetric` result object.
+Although the `overall` and `by_groups` properties will work fine, the `differences()` and `ratios()` methods may not.
+After all, what does "take the ratio of two confusion matrices" even mean?
+We should try to trap these cases, and throw a meaningful exception (rather than propagating whatever exception happens to emerge from the underlying libraries).
+Since we know that `differences()` and `ratios()` will only work when the metric has produced scalar results, this should be a straightforward test.
+
+## Pitfalls
+
+There are some potential pitfalls which could trap the unwary.
+
+The biggest of these are related to missing classes in the subgroups.
+To take an extreme case, suppose that males were always being predicted classes A or B, while females were always predicted classes C or D.
+The user could request precision scores, but the results would not really be comparable between the two groups.
+With intersections of sensitive features, cases like this become more likely.
+
+Metrics in SciKit-Learn usually have arguments such as `pos_label=` and `labels=` to allow the user to specify the expected labels, and adjust their behaviour accordingly.
+However, we do not require that users stick to the metrics defined in SciKit-Learn.
```

Comment:
but we can require people to use metrics which allow `labels=` for these cases. Or document the pitfalls if the underlying metric doesn't support this.

Reply:
I'm not quite sure what you mean. As I say in the next paragraph, if we're providing convenience strings for the functions, then we can perhaps require them to pass `labels=` when the metric function supports it. I suppose we might be able to figure out if an actual _function_ passed in came from SciKit-Learn, but that makes us even more fragile to changes in SciKit-Learn (and I'm already a bit leery in the user-specifies-string case).

If you mean inspecting the function's arguments (I assume that Python can do that sort of reflection) for one called `labels=` then that's horribly fragile. We don't actually know if we've even been given a classification function in the first place.

Reply:
My point is that here the text says "_ However, we do not require that users stick to the metrics defined in SciKit-Learn._" implying that if the underlying function supports `labels` we're fine.

There are two ways you can check if the underlying function supports it, one is to call it passing `labels` and if you get a `KeyError` (I think) you know there's been an issue and you raise an appropriate error. The other is to check the signature with `inspect`.

I'd say it's reasonable to expect certain functionalities from the underlying function in the case of classification.

My point is that we don't really need to know or check if it's a sklearn metric or not, we just need to check if it supports what we need.

Reply:
I agree that a robust classification metric should support some means of being told which classes are supposed to be present. I'm concerned that not all metrics will use `labels=` for that purpose. They might use some other name, in which case we will have lulled users into a false sense of security. They might also use `labels=` for some other purpose as well, which would be an even bigger mess.